### PR TITLE
Use strongerrors in provider implementations.

### PR DESCRIPTION
**Pull request by cpuguy83**
This ensures that we can catch certain types of errors from providers
and handle accordingly in the core. There is still more to do here to
improve that but this resolves an immediate need to know why a Delete
failed.

vic provider was not updated since I could not figure out where to get
this information.

### CODE REVIEW

**Comment by robbiezhang** [comment]
@cpuguy83 Nice job. I'll take a look.

File: providers/azure/client/api/api.go [hunk 1]
```diff
@@ -56,12 +61,20 @@ func CheckResponse(res *http.Response) error {
 			}
 			jerr.Error.Body = string(slurp)
 			jerr.Error.URL = res.Request.URL.String()
-			return jerr.Error
+			e = jerr.Error
+		}
+	} else {
+		e = &Error{
+			StatusCode: res.StatusCode,
+			Body:       string(slurp),
```
Comment (robbiezhang) [inline providers/azure/client/api/api.go]: old issue: slurp is nill here. Shall we use res.Status instead?

File: providers/azure/client/api/api.go [hunk 2]
```diff
@@ -56,12 +61,20 @@ func CheckResponse(res *http.Response) error {
...
+			Header:     res.Header,
 		}
 	}
-	return &Error{
-		StatusCode: res.StatusCode,
-		Body:       string(slurp),
-		Header:     res.Header,
+
+	switch e.StatusCode {
```
Comment (robbiezhang) [inline providers/azure/client/api/api.go]: wondering whether we should implement it in the aci provider like what you did for the other providers.

Comment (cpuguy83) [reply providers/azure/client/api/api.go]: Yeah, I definitely cheated here.

File: providers/azure/client/api/api.go (hunk 1 above)
Comment (cpuguy83) [reply providers/azure/client/api/api.go]: Fixed

File: providers/azure/client/api/api.go (hunk 2 above)
Comment (cpuguy83) [reply providers/azure/client/api/api.go]: Moved this handling into the provider.

**Pull request merged**